debug Api NodeJS

DenisArger opened the pull request:
Данный ПР представляет собой финальную стадию разработки api с использованием node.js.
Для удобства разработки фунционал разбит на модули:
**- generatorId.js:** получение уникального с использованием внешнего модуля nanoid
**- supabase.js:** взаимодействие БД (добавление  новой записи,  получение объекта json) с использованием внешнего  модуля  @supabase/supabase-js
**- express.js**: api  ожидания get и post запросов  с использованием внешнего модуля express

DenisArger committed:
пакетный менеджер yarn
добавлена обработка отсутствия объекта с указанным id
изменен файл gitignore

DenisArger committed:
Изменения с прошлого коммита

DenisArger committed:
Переменные изменены на const

foxprogs commented on .yarn/releases/.gitignore:
```diff
@@ -0,0 +1 @@
+/yarn-3.5.0.cjs
```
удали совсем эту папку пожалуйста

DenisArger replied:
ок

foxprogs commented on geheratorId.js:
```diff
@@ -0,0 +1,9 @@
+    const nanoid = customAlphabet(alphanumeric, 6);
+    return nanoid();
+  },
+};
```
название файла с опечаткой, geNerator

DenisArger replied:
точно)

foxprogs commented on express.js:
```diff
@@ -0,0 +1,38 @@
+const express = require("express");
+const { getJsonData } = require("./supebase.js");
+const { addJsonData } = require("./supebase.js");
```
можно объеденить

DenisArger replied:
готово

foxprogs commented on express.js:
```diff
@@ -0,0 +1,38 @@
+});
+
+// создаем парсер для данных application/x-www-form-urlencoded
+const urlencodedParser = express.urlencoded({ extended: true, limit: "1mb" });
```
это же для картинок вроде

DenisArger replied:
Уточнил. Это метод, встроенный в express для распознавания входящего объекта запроса в виде строк или массивов. 
Получается не мой случай.

DenisArger replied:
Вынужден оставить. Без него ошибка в post запросе.

foxprogs commented on package.json:
```diff
@@ -0,0 +1,10 @@
+    "dotenv": "^16.0.3",
+    "express": "4.17.0",
+    "nanoid": "3.3.6",
+    "nanoid-dictionary": "^4.3.0"
```
можно без него

DenisArger replied:
удалить этот файл?

foxprogs replied:
да, библиотеку удалить, она по сути просто строку тебе отдает. Вот тут дока для третьей версии, как кастомный алфавит добавить
https://github.com/ai/nanoid/tree/v3#custom-alphabet-or-size

DenisArger replied:
Понял, спасибо.

DenisArger committed:
Исправлено имя файла generatorId
Объединено подключение generatorId

DenisArger committed:
Удален файл .yarn/releases/.gitignore

foxprogs commented on .gitignore:
```diff
@@ -128,3 +127,5 @@ dist
 .yarn/build-state.yml
 .yarn/install-state.gz
 .pnp.*
+/yarn.lock
```
это вроде не надо, можно убрать

foxprogs commented on .gitignore:
```diff
@@ -128,3 +127,5 @@ dist
 .yarn/install-state.gz
 .pnp.*
+/yarn.lock
+/.editorconfig
```
это лучше оставлять в репо, и можно писать без / как в примерах выше

foxprogs commented on supebase.js:
```diff
@@ -0,0 +1,54 @@
+
+const supabaseApp = supabase.createClient(supabaseUrl, supabaseKey);
+
+const { getId } = require("./generatorId");
```
старайся все require вверху файла добавить

foxprogs commented on supebase.js:
```diff
@@ -0,0 +1,54 @@
+          throw error;
+        }
+      } else {
+        throw new NotJsonException();
```
https://learn.javascript.ru/try-catch#generatsiya-sobstvennyh-oshibok
если хочешь кастомную ошибку то вот тут прочти

DenisArger replied:
Отлично! Спасибо!

foxprogs commented on supebase.js:
```diff
@@ -0,0 +1,54 @@
+    const { data, error } = await supabaseApp
+      .from("jsons_tb")
+      .select("data")
+      .eq("id", id);
```
надо дописать так чтобы он возвращал одну запись а не массив

DenisArger replied:
хорошо, подправлю

foxprogs replied:
я имел ввиду вот это https://supabase.com/docs/reference/javascript/single

DenisArger replied:
готово

DenisArger committed:
Подключение generatorId перенесено вверх файла
Подправлен файл gitignore

DenisArger committed:
Удалено использование библиотеки nanoid-dictionary

DenisArger committed:
Получение одиночного значения перенесено в метод
getJsonData

foxprogs commented on supebase.js:
```diff
@@ -0,0 +1,57 @@
+require("dotenv").config();
```
надо поменять название файла на supabase.js

DenisArger replied:
готово

DenisArger committed:
Изменено название файла supeBase на supaBase

foxprogs commented on express.js:
```diff
@@ -0,0 +1,33 @@
+    return response.sendStatus(400);
+  }
+  addJsonData(request.body);
+  response.send(`${request.body}`);
```
лучше вернуть json с состоянием, если успешно сохранено то вернуть айди, если ошибка то написать какая именно

foxprogs commented on express.js:
```diff
@@ -0,0 +1,33 @@
+  if (!request.body) {
+    return response.sendStatus(400);
+  }
+  addJsonData(request.body);
```
надо результат же как-то обработать. Ошибка или сохранилось. И соответстенно вернуть

foxprogs commented on supebase.js:
```diff
@@ -0,0 +1,57 @@
+
+async function addJsonData(data) {
+  try {
+    try {
```
не совсем понял почему два блока, ты же во внyтреннем catch не выкидываешь исключение throw чтобы его внешний обработал. Я бы сделал один, но по названию исключения уже обрабатывал. Ну и обычного console.log тут н едостаточно. Надо чтобы возвращал, успех или ошибка.

foxprogs commented on supebase.js:
```diff
@@ -0,0 +1,57 @@
+        const id = getId();
+        const { error } = await supabaseApp
+          .from("jsons_tb")
+          .insert([{ id: id, data: data }]);
```
тут можно использовать сокращенную запись `.insert([{ id, data }]);`

foxprogs commented on index.html:
```diff
@@ -0,0 +1,18 @@
+  <h1>Введите данные</h1>
+  <form method="post">
+    <label>ID</label><br>
+    <input type="text" name="idJson" /><br><br>
```
давай в примере сделаем textarea куда надо вбить json и сделай чтобы там уже по умолчанию был какой-то.
Ну и считай что этот `index.html` это типа твоей документации. По этому я бы тут расписал как этим пользоваться и привел примеры.
В задаче у тебя должен быть POST запрос к `/generate` а не к `/`, это надо исправить
Тут в примере показать что вот мы отправляем json, если будет кривой то такая-то ошибка.
Если нормальный то вернулся вот айди, если теперь вот тут вбить айди то мы получим наш объект обратно.

DenisArger committed:
Функция добавления нового объекта возращает json
со статусом

DenisArger committed:
Реализована обработка исключений и возвращение
статуса ошибок

DenisArger committed:
Исправлен путь с post запросу

DenisArger committed:
Добавлен файл post для отправки post запроса

DenisArger committed:
Добавлен небольшой мануал по отправке запросов

DenisArger committed:
Добавлена обработка главной страницы

foxprogs commented on express.js:
```diff
@@ -0,0 +1,48 @@
+  const id = request.params.id;
+
+  if (id.length > 6) {
+    response.json({ error: "error" });
```
Что именно за ошибка. И код ошибки не 200 нужен а другой

foxprogs commented on express.js:
```diff
@@ -0,0 +1,48 @@
+});
+
+app.get("/get", function (request, response) {
+  response.send(
```
Это апи, он не должен просто текст возвращать, нужны http код ответа и json с данными

DenisArger replied:
уберу этот момент

foxprogs commented on express.js:
```diff
@@ -0,0 +1,48 @@
+  response.sendFile(__dirname + "/index.html");
+});
+
+app.get("/get", function (request, response) {
```
Это можно отдельно не делать, просто где у тебя айди, там укажи что он опциональный, и выдай ошибку

DenisArger replied:
хорошо, так и сделаю

foxprogs commented on index.html:
```diff
@@ -0,0 +1,24 @@
+<html>
+
+<head>
+  <title>Регистрация</title>
```
Заголовок надо исправить

foxprogs commented on express.js:
```diff
@@ -0,0 +1,48 @@
+      response.json(data);
+    });
+  } catch (error) {
+    response.json(error);
```
Вот тут ты мог бы тоже статус отправлять, а то по дефолту 200 вроде. У тебя же приходит статус сюда?

DenisArger replied:
От supebase приходит статус 201 при успешном добавлении объекта.
У меня функция addJsonData возвращает статусы. В post запрос приходят ответы в формате json с успешным ответом или с кодом ошибки.

foxprogs commented on express.js:
```diff
@@ -0,0 +1,48 @@
+    return response.sendStatus(400);
+  }
+  try {
+    addJsonData(request.body).then(function (data) {
```
Если ты промисы используешь то тогда можно в трайкэч не оборачивать а там после then использовать catch блок

foxprogs commented on express.js:
```diff
@@ -0,0 +1,48 @@
+    response.json({ error: "error" });
+  } else {
+    getJsonData(id).then(function (data) {
+      response.json(data);
```
И здесь добавить надо тоже http код для ошибки и успеха

DenisArger replied:
у меня  getJsonData возвращает json  с данными или код ошибки. 
Что-то еще нужно добавить?

foxprogs commented on express.js:
```diff
@@ -0,0 +1,48 @@
+  );
+});
+
+app.get("/generate", function (request, response) {
```
Эту страницу убери и попробуй сделать документацию по твоему апи на одной странице index

DenisArger committed:
Исправлен title главной страницы

DenisArger committed:
Добавлена обработка ошибки при неверном id
Исправлен блок catch для промиса

DenisArger committed:
Реализован get запрос с обработчиком ошибок

DenisArger committed:
Добавлен post запрос c обработком ошибок

DenisArger committed:
Добавлен скрипт для взаимодействия с документацией

DenisArger committed:
Добавлена документация взаимодействия с api

DenisArger committed:
каталог vs code  добавоен в исключения

DenisArger committed:
Исправлено сообщение при неправильном вводе json
данных

DenisArger committed:
Убрана лишняя информация

DenisArger committed:
Убрана ненужная информация
Селекторы перенесены внутрь функций
Добавлена проверка на пустые данныеd textarea

DenisArger committed:
Удалил ненужный файл post.html

DenisArger committed:
Исправлен оператор сравнения

DenisArger committed:
Добавлена обработка неправильного формирования
 json данных в textarea

DenisArger committed:
Merge pull request #2 from tt-dev-23/push

слияние с документацией
